File: fanotify/file-watcher/watcher.c
#define _GNU_SOURCE /* for O_LARGEFILE */
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <limits.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/fanotify.h>

#include <poll.h>

#include "watcher.h"

static struct pollfd *fds = NULL;
static nfds_t nfds = 0;

static int handle_event(int fd, const struct fanotify_event_metadata *metadata,
                        ssize_t len) {
	int ret = EXIT_SUCCESS;
	char *path = NULL;
        char *prog_path = NULL;
        ssize_t path_len;
        char procfd_path[PATH_MAX];
	struct fanotify_response response;

	/* filepath */
        memset(procfd_path, 0, PATH_MAX);
        snprintf(procfd_path, sizeof(procfd_path), "/proc/self/fd/%d",
                 metadata->fd);
        path = (char*)calloc(PATH_MAX, sizeof(char));
	path_len = readlink(procfd_path, path, PATH_MAX - 1);
	if (path_len == -1) {
		perror("readlink");
	}
	path[path_len] = '\0';
	printf("PATH: %s\n", path);

	/* program name */
	memset(procfd_path, 0, PATH_MAX);
        snprintf(procfd_path, sizeof(procfd_path), "/proc/%d/exe",
                 metadata->pid);
        prog_path = (char *)calloc(PATH_MAX, sizeof(char));
        path_len = readlink(procfd_path, prog_path, PATH_MAX - 1);
        if (path_len == -1) {
		perror("readlink");
	}
	prog_path[path_len] = '\0';
	printf("PROG PATH: %d -- %s\n", metadata->pid, prog_path);

        if (metadata->mask & FAN_OPEN_EXEC_PERM) { /* OPEN_EXEC_PERM  */
		printf("FAN_OPEN_EXEC_PERM: '%s' <-- '%s'\n", path, prog_path);
		response.fd = metadata->fd;
		response.response = FAN_ALLOW;
		write(fd, &response, sizeof(response));
        } else if (metadata->mask & FAN_CLOSE_WRITE) {/* CLOSE_WRITE */
		printf("FAN_CLOSE_WRITE: '%s' <-- '%s'\n", path, prog_path);
        } else if (metadata->mask & FAN_MODIFY) {/* MODIFY */
		printf("FAN_MODIFY: '%s' <-- '%s'\n", path, prog_path);
	}

	free(path);
	free(prog_path);
	close(metadata->fd);
	return ret;
}

#define MAX_BUF 256

static int read_event(int fd) {
	const struct fanotify_event_metadata *metadata = NULL;
	struct fanotify_event_metadata buf[MAX_BUF];
	ssize_t len;

	len = read(fd, buf, sizeof(buf));
	if (len == -1 && errno != EAGAIN) {
		perror("read");
		return -11;
	}

	/* check if end of available data reached. */
	if (len <= 0)
		return -12;

	/* point to the first event in the buffer */
	metadata = buf;

	/* loop over all events in the buffer */
	while (FAN_EVENT_OK(metadata, len)) {
		/* check that runtime and compile-time structures match */
		if (metadata->vers != FANOTIFY_METADATA_VERSION) {
			fprintf(stderr, "Mismatch of fanotify metadata version.\n");
			return -13;
		}

		/* metadata->fd contains either FAN_NOFD, indicating a
		   queue overflow, or a file descriptor (a nonnegative
		   integer). Here, we simply ignore queue overflow. */
		if (metadata->fd < 0) {
			continue;
		}

		// TODO(jouyouyun): thread
		if (handle_event(fd, metadata, len) == EXIT_FAILURE)
			return -14;

		metadata = FAN_EVENT_NEXT(metadata, len);
	}

	return 0;
}

static int
setup_fanotify(const char *filename, unsigned int init_flags,
	       unsigned int mark_flags, unsigned int event_flags)
{
	int fd = 0;

	fd = fanotify_init(init_flags, O_RDONLY|O_LARGEFILE);
	if (fd == -1) {
		perror("fanotify_init");
		return -1;
	}

	if (fanotify_mark(fd, mark_flags, event_flags, AT_FDCWD, filename) == -1) {
		perror("fanotify_mark");
		close(fd);
		return -2;
	}
	
	return fd;
}

static int setup_poll(int fd)
{
	struct pollfd *tmp = NULL;
	
	tmp = realloc(fds, sizeof(struct pollfd) * (nfds + 1));
	if (!tmp) {
		perror("realloc");
		return -3;
	}

	fds = tmp;
	fds[nfds].fd = fd;
	fds[nfds].events = POLLIN;
	nfds++;

	return 0;
}

int watch_file(const char *filename, unsigned int init_flags,
               unsigned int mark_flags, unsigned int event_flags)
{
	int fd = 0;
	int ret = 0;

	fd = setup_fanotify(filename, init_flags, mark_flags, event_flags);
	if (fd < 0) {
		return fd;
	}

        ret = setup_poll(fd);
	if (ret < 0) {
		close(fd);
		return ret;
	}
	
        return fd;
}

int watch_remove(int fd)
{
	int i = 0;
	int len = sizeof(struct pollfd);
	struct pollfd *tmp = NULL;

	// TODO(jouyouyun): lock
	for (; i < nfds; i++) {
		if (fd == fds[i].fd) {
			break;
		}
	}
	if (i == nfds)
		return -31;

	tmp = calloc(nfds - 1, len);
	memcpy(tmp, fds, len * i);
	memcpy(tmp+i, fds+i+1, len * (nfds - i));

	free(fds);
	fds = tmp;
	nfds -= 1;
	close(fd);

	return 0;
}

int poll_event() {
	int i = 0;
	int ret = 0;
	int num = 0;
	int timeout = 3;
	static int loop = 0;

	if (loop)
		return 0;

	if (nfds < 1)
		return -21;

	loop = 1;
	while (1) {
		num = poll(fds, nfds, timeout);
		if (num == -1) {
			perror("poll");
			goto error;
		}
		if (num < 1)
			goto error;

		for (i = 0; i < nfds; i++) {
			if (fds[i].revents & POLLIN) {
				ret = read_event(fds[i].fd);
				if (ret < 0)
					fprintf(stderr, "Failed to handle event: %d\n", ret);
			}
		}
		continue;

	error:
		usleep(500 * 1000); // 500ms
	}
}

File: fanotify/mount-watcher/watcher.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"os"
	"sync"

	"golang.org/x/sys/unix"
)

type fileInfo struct {
	fd     int
	fp     *os.File
	reader *bufio.Reader
}

type EventInfo struct {
	Filename string
	Program  string
	File     *fileInfo
	unix.FanotifyEventMetadata
}

type Watcher struct {
	locker  sync.RWMutex
	fileSet map[string]*fileInfo

	Event chan EventInfo
}

func (ev *EventInfo) MaskMatch(mask uint64) bool {
	return (ev.Mask&mask == mask)
}

func (ev *EventInfo) Close() error {
	return unix.Close(int(ev.Fd))
}

func (ev *EventInfo) Allow() error {
	return binary.Write(ev.File.fp, binary.LittleEndian, &unix.FanotifyResponse{
		Fd:       ev.Fd,
		Response: unix.FAN_ALLOW,
	})
}

func (ev *EventInfo) Deny() error {
	return binary.Write(ev.File.fp, binary.LittleEndian, &unix.FanotifyResponse{
		Fd:       ev.Fd,
		Response: unix.FAN_DENY,
	})
}

func NewWatcher() (*Watcher, error) {
	var w = &Watcher{
		fileSet: make(map[string]*fileInfo),
		Event:   make(chan EventInfo, 1024),
	}
	return w, nil
}

func (w *Watcher) Watch(filename string, initFlags, markFlags, eventFlags uint) error {
	if w.isFileExists(filename) {
		return fmt.Errorf("Has exists: %s", filename)
	}

	fd, err := unix.FanotifyInit(initFlags, unix.O_RDONLY|unix.O_LARGEFILE)
	if err != nil {
		return err
	}
	err = unix.FanotifyMark(fd, markFlags, uint64(eventFlags), unix.AT_FDCWD, filename)
	if err != nil {
		_ = unix.Close(fd)
		return err
	}

	w.addFile(filename, fd)
	go w.loop(filename)
	return nil
}

func (w *Watcher) Remove(filename string) error {
	if !w.isFileExists(filename) {
		return fmt.Errorf("Not found: %s", filename)
	}

	fi := w.removeFile(filename)
	fi.fp.Close()
	return nil
}

func (w *Watcher) Close() {
	w.locker.Lock()
	defer w.locker.Unlock()
	for _, v := range w.fileSet {
		v.fp.Close()
	}
	close(w.Event)
}

func (w *Watcher) loop(filename string) error {
	var err error
	fi := w.getFileInfo(filename)
	for {
		var event unix.FanotifyEventMetadata
		err = binary.Read(fi.reader, binary.LittleEndian, &event)
		if err != nil && err != unix.EAGAIN {
			fmt.Println("Failed to read:", err)
			break
		}

		if event.Vers != unix.FANOTIFY_METADATA_VERSION {
			err = fmt.Errorf("Mismatch version")
			break
		}

		file, _ := os.Readlink(fmt.Sprintf("/proc/self/fd/%d", event.Fd))
		progPath, _ := os.Readlink(fmt.Sprintf("/proc/%d/exe", event.Pid))
		w.Event <- EventInfo{
			Filename:              file,
			Program:               progPath,
			File:                  fi,
			FanotifyEventMetadata: event,
		}
	}

	w.Remove(filename)
	return err
}

func (w *Watcher) isFileExists(filename string) bool {
	w.locker.RLock()
	_, ok := w.fileSet[filename]
	w.locker.RUnlock()
	return ok
}

func (w *Watcher) addFile(filename string, fd int) {
	fp := os.NewFile(uintptr(fd), "")
	w.locker.Lock()
	w.fileSet[filename] = &fileInfo{
		fd:     fd,
		fp:     fp,
		reader: bufio.NewReader(fp),
	}
	w.locker.Unlock()
}

func (w *Watcher) removeFile(filename string) *fileInfo {
	w.locker.Lock()
	fi := w.fileSet[filename]
	delete(w.fileSet, filename)
	w.locker.Unlock()
	return fi
}

func (w *Watcher) getFileInfo(filename string) *fileInfo {
	w.locker.RLock()
	info := w.fileSet[filename]
	w.locker.RUnlock()
	return info
}
